File: challengeFive/arduino/car_example.py
import time

# basically just proportionality-control algorithm since Ki=Kd=0
Kp = 1
Ki = 0
Kd = 0

# desired output (go straight)
setpoint = 90

# initial position: left side of hallway (needs to turn RIGHT)
lidar_L = 200
lidar_R = 600
delta = lidar_L - lidar_R

# normalize from -1 to +1  (saying min/max is 25-meter delta)
delta_N = 2*((delta + 2500)/(5000)) - 1

# normalize from -90 to +90
delta_90 = 90 * delta_N

# set initial crawler lidar input (90 +- 90 = 0 to 180)
process_feedback = 90 - delta_90
print("Initial crawler input: ", process_feedback)
if (process_feedback < 90):
	print("(LEFT turn required from initial position", end="")
elif (process_feedback > 90):
	print("(RIGHT turn required from initial position", end="")
print(" in order to move process_feedback toward setpoint.)\n\n")	
time.sleep(2)


# poll period in seconds (will be ~0.2 on car)
dt = 2

# main control loop
previous_error = setpoint - process_feedback
integral = 0
while (1):
	time.sleep(dt)
	# print goal and actual output first
	print("setpoint: ", setpoint)
	print("process feedback: ", process_feedback)
	error = setpoint - process_feedback
	integral = integral + (error*dt)
	derivative = (error - previous_error)/dt
	Output = (Kp*error) + (Ki*integral) + (Kd*derivative)
	previous_error = error
	print("error:", error)
	#print("integral:", integral)
	#print("derivative: ", derivative)
	print("Output: ", Output)
	print("previous_error: ", previous_error)
	print("\n")
	# input being changed according to output
	process_feedback = process_feedback + Output 
	
		
	
	
	
	

File: challengeFive/arduino/car_example.c


#include <stdio.h>
#include <unistd.h>

/* GLOBAL VARIABLES */

// basically just proportionality-control algorithm since Ki=Kd=0
const double Kp = 1;
const double Ki = 0;
const double Kd = 0;

// desired wheel output (go straight)
const double setpoint = 90; // may change to int, but not sure about concatenation issues

// initial position for testing (car is closer to right wall, needs to move LEFT)
const int lidar_L = 20;
const int lidar_R = 600;
const double delta = lidar_L - lidar_R;

// normalize delta to ±90 (estimating max difference is ±25-meters)
const double delta_90 = 90*(2*((delta + 2500)/(5000)) - 1);

// crawler output
double process_feedback;

// poll period in microseconds (will actually be ~0.2 seconds on car)
const unsigned int dt = 1000000;


// for testing purposes	
void setup() {
	// set initial crawler lidar input (±90 --> 0 to 180)
	process_feedback = 90 - delta_90;
	printf("Left distance from wall: %dcm\n", lidar_L);
	printf("Right distance from wall: %dcm\n", lidar_R);
	if (lidar_L > lidar_R) {
		printf("Starting car closer to right wall.\n");
	}
	else if ( lidar_L < lidar_R) {
		printf("Starting car closer to left wall.\n");
	}
	else {
		printf("Starting car evenly between walls.\n");
	}
	//printf("Initial crawler input: %f\n", process_feedback);
	if (process_feedback < 90) {
		printf("LEFT turn required from initial position in order to ");
		printf("move process_feedback toward setpoint.\n");
	}
	else if (process_feedback > 90) {
		printf("RIGHT turn required from initial position in order to ");
		printf("move process_feedback toward setpoint.\n");	
	}
	else if (process_feedback == 90) {
		printf("Going STRAIGHT already! (process_feedback == setpoint)\n");		
	}
	printf("Starting...\n\n");
}
	
	
void PID_Control(double previous_error, double error, double integral, 
				 double derivative, double output) {
	usleep(dt);	
	printf("setpoint: %f\n", setpoint);
	printf("process feedback: %f\n", process_feedback);
	error = setpoint - process_feedback;
	integral = integral + (error * dt);
	derivative = (error - previous_error) / dt;
	output = (Kp * error) + (Ki * integral) + (Kd * derivative);
	previous_error = error;
	printf("error: %f\n", error);
	//printf("integral: %d\n", integral);
	//printf("derivative: %d\n", derivative);
	printf("PID output: %f\n", output);  // for full output: %.20lf
	if ((output >= -0.000001) && (output <= 0.000001)){ // (supposedly bad practice to use == with floats,
		printf("Going STRAIGHT.\n");        // but when it shows -0.000000 it's actually less than 0) 
	}
	else if (output > 0.000001) {
		printf("Turning LEFT.\n");
	}
	else if (output < -0.000001) {
		printf("Turning RIGHT.\n");
	}
	//printf("previous_error: %d\n", previous_error);
	printf("\n");
	
	// artifical parameter so it doesn't think it corrected input perfectly in just 1 iteration
	double scaling = 0.99; // set to 1 for 1-iteration correction
	// input being changed according to output
	process_feedback += scaling * output;
}	

	

int main() {
	// initial conditions according to lidar test parameters
	setup();
	
	// pause
	unsigned int sleep_μS = 1250000;
	usleep(sleep_μS);
	
	// initial variables
	double previous_error = setpoint - process_feedback;
	double integral = 0;
	double error, derivative, output;
	
	// main loop
	while (1) 
		PID_Control(previous_error, error, integral, derivative, output);	
	
}
		


	
	
	
	
